pkg/tools/serialize/serialize.go

```diff
@@ -19,59 +19,50 @@
 // will be executed in order
 package serialize
 
-import "runtime"
+import (
+	"sync"
+	"sync/atomic"
+)
 
-// Executor - allows one at a time in order execution of func()s.  func()s can be queued Asynchronously or
-// Synchronously
-type Executor interface {
-	AsyncExec(func())
-	SyncExec(func())
-}
+const (
+	channelSize = 100
+)
 
-type executor struct {
-	execCh      chan func()
-	finalizedCh chan struct{}
+// Executor - a struct that can be used to guarantee exclusive, in order execution of functions.
+type Executor struct {
+	execCh chan func()
+	init   sync.Once
+	count  int32
 }
 
-// NewExecutor - returns a new Executor
+// NewExecutor - provides a new Executor
+// Deprecated: Please just used Executor{} in the future.  The zero value of Executor works just fine.
 func NewExecutor() Executor {
-	rv := &executor{
-		execCh:      make(chan func(), 100),
-		finalizedCh: make(chan struct{}),
-	}
-	go eventLoop(rv.execCh, rv.finalizedCh)
-	runtime.SetFinalizer(rv, func(f *executor) {
-		close(f.finalizedCh)
-	})
-	return rv
-}
-
-// eventLoop is intentionally written as a separate function passed channels directly
-// because if we make it a receiver on the executor{} then it will prevent garbage collection of the
-// executor.
-func eventLoop(execCh chan func(), finalizedCh chan struct{}) {
-	for {
-		select {
-		case exec := <-execCh:
-			exec()
-			continue
-		case <-finalizedCh:
-		}
-		break
-	}
-}
-
-// AsyncExec - queues exec for execution and returns without waiting for exec() to run
-func (t *executor) AsyncExec(exec func()) {
-	t.execCh <- exec
+	return Executor{}
 }
 
-// SyncExec - queues exec for execution and returns *after* exec() has run
-func (t *executor) SyncExec(exec func()) {
-	done := make(chan struct{})
-	t.execCh <- func() {
-		exec()
+// AsyncExec - guarantees f() will be executed Exclusively and in the Order submitted.
+//        It immediately returns a channel that will be closed when f() has completed execution.
+func (e *Executor) AsyncExec(f func()) (done chan struct{}) {
+	// Initialize *once*
+	e.init.Do(func() {
+		e.execCh = make(chan func(), channelSize)
+	})
+	done = make(chan struct{})
+	e.execCh <- func() {
```

Comment:
Should we move e.execCh <- after we do increment? It would be similar to

var wg sync.WaitGroup
for i := 0; i < b.N; i++ {
 	wg.Add(1)
 	go func() { wg.Done() }
 }
wg.Wait()

Reply:
I don't think so.  e.count represents the *unexecuted* func()s we have.  Just taking it off of the channel doesn't means its executed yet.  So we aren't really at e.count == 0 until g() has finished executing.  If someone adds a func() to the queue *while* its executing... then when we atomically deprecate e.count, it won't be zero, and we will go around the loop again :)

Reply:
Wait... totally misread... retract the last comment... one moment

Reply:
OK... we could increment before e.execCh <- , but it increases the complexity of the code, and in running the Benchmarks doesn't make it any faster:

```go
	count := atomic.AddInt32(&e.count, 1)
	e.execCh <- func() {
		f()
		close(done)
	}
	// Start go routine if we don't have one
	if count == 1 {
		go func() {
			for g := range e.execCh {
				g()
				if atomic.AddInt32(&e.count, -1) == 0 {
					return
				}
			}
		}()
	}
	return done
```

Reply:
Since there doesn't *appear* to be a perf benefit, and both seem to be correct, shall we go with the simpler code version?

Reply:
Actually I'e mean following change:
```go
if atomic.AddInt32(&e.count, 1) == 1 {
	go func() {
		for g := range e.execCh {
			g()
			if atomic.AddInt32(&e.count, -1) == 0 {
				return
			}
		}
	}()
}
e.execCh <- func() {
	f()
	close(done)
}
```
In my experiments all variants looks working fine. So I'm ok having any of them. Moving e.execCh <- is few ms better on my env, not sure it is make any seance :-)

Reply:
I just pushed an update.  You are right, this is about 5% faster (at least when I move the channel allocate after as well) :)